File: exercise/5/solution2.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.length() == 0) {
            return "";
        }
        vector<vector<int>> f(s.length(), vector<int>(s.length(), 0));
		int max_length = 0, left = 0;
		for (int i = 0; i < s.length(); i++) {
			for (int j = i; j >= 0; j--) {
				if (s[j] == s[i] && (i - j < 2 || f[j+1][i-1] == 1)) {
					f[j][i] = 1;
					if (max_length < i - j) {
						max_length = i - j;
						left = j;
					}
				}
			}
		}
		return s.substr(left, max_length);
    }
};

File: exercise/5/solution2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        is_palind = [[False for __ in range(len(s))] for _ in range(len(s))]
        max_length = 1
        left = 0
        for i in range(len(s)):
            j = i
            # from j to i(include j and i)
            while j >= 0:
                if s[i] == s[j] and (i - j < 2 or is_palind[i-1][j+1]):
                    is_palind[i][j] = True
                    if i - j + 1 > max_length:
                        max_length = i - j + 1
                        left = j
                j -= 1
        return s[left:left+max_length]
